### PR TITLE
Fermi Net implementation for qmctorch

### PR DESCRIPTION
**Pull request by Rik-Breebaart**
A implementation of the FermiNet from [arXiv:1909.02487].

### CODE REVIEW

File: qmctorch/wavefunction/FermiNet/FermiNet.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+        npParamCount = np.prod(param.data.shape)
+        numParams += npParamCount
+        if param.requires_grad:
+            numTrainable += npParamCount
+    return numParams, numTrainable
+
+#from https://discuss.pytorch.org/t/when-should-i-use-nn-modulelist-and-when-should-i-use-nn-sequential/5463/12
+
+
+# test the intermediate layers .
```
[the remark this answers is not in the document]
Comment (NicoRenaud) [reply qmctorch/wavefunction/FermiNet/FermiNet.py:256]: now if you don't use this file all that doesn't matter of course

File: qmctorch/wavefunction/FermiNet/FermiNet.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
+import torch
```
[the remark this answers is not in the document]
Comment (Rik-Breebaart) [reply qmctorch/wavefunction/FermiNet/FermiNet.py:1]: No, I don't use it anywhere, I will remove it fromt the repo.

File: qmctorch/wavefunction/wf_FermiNet.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            pos (float): electron position Nbatch x [Nelec x Ndim]      
+        
+        '''
+        Nbatch = pos.shape[0]
+
+        # Go through the intermediate layers 
+        h_L_i, h_L_ij = self.Intermediate_layer(pos)
+        # Create the determinants 
+        determinant = torch.zeros(Nbatch, self.K_determinants)
+        for k in range(self.K_determinants):
```
[the remark this answers is not in the document]
Comment (Rik-Breebaart) [reply qmctorch/wavefunction/wf_FermiNet.py:88]: I first used a meshgrid to get the desired orbitals and electrons indices in de slatter determinant, however as the orbitals where nn.layers the indexing did not work. I will look into the method in slater_pooling.

File: qmctorch/wavefunction/wf_FermiNet.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        """
+        d = []
+        for iat in range(self.natom):
+            at = self.atoms[iat]
+            xyz = self.ao.atom_coords[iat,
+                                      :].detach().numpy().tolist()
+            d.append((at, xyz))
+        return d
+
+    def pdf(self, pos):
```
[the remark this answers is not in the document]
Comment (Rik-Breebaart) [reply qmctorch/wavefunction/wf_FermiNet.py:199]: Certainly. I will change that.

File: qmctorch/wavefunction/wf_FermiNet.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        if not torch.cuda.is_available and self.wf.cuda:
+            raise ValueError('Cuda not available, use cuda=False')
+
+        # number of atoms
+        self.mol = mol
+        self.atoms = mol.atoms
+        self.natom = mol.natom
+        self.nup = mol.nup
+        self.ndown = mol.ndown
+        self.nelecetrons = self.nup + self.ndown
```
[the remark this answers is not in the document]
Comment (Rik-Breebaart) [reply qmctorch/wavefunction/wf_FermiNet.py:41]: yes, I will change that and clean up to take the variables from mol.

File: qmctorch/wavefunction/wf_FermiNet.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        self.ndim =3
+        
+        # hyperparameters of network
+        self.hidden_nodes_e = hidden_nodes_e
+        self.hidden_nodes_ee = hidden_nodes_ee
+        self.K_determinants = K_determinants
```
[the remark this answers is not in the document]
Comment (Rik-Breebaart) [reply qmctorch/wavefunction/wf_FermiNet.py:47]: In principle since the orbitals are dependent on not only one electron but all others aswell the network should be able to fit any antisymmetric function. This is explained in the Supplementary Information of the Foulkes's paper. However this is dependent on the architecture of the network and requires non-continues functions which are directly learnable by the FermiNet. Therefore multiple determinants are required to reach high accuracy, but this is not explained in relation to the configs.

[the remark this answers is not in the document]
Comment (Rik-Breebaart) [reply qmctorch/wavefunction/wf_FermiNet.py:47]: *non-continues functions which are not directly learnable by the FermiNet

**Comment by NicoRenaud** [comment]
It seems that you have reverted the last merges to the master and that somehow the commit history is confused. If you look at the files changed all the logs have disappeared.  Could you merge the master back into your branch ? Save your files about the FermiNet somewhere first ;)

**Comment by Rik-Breebaart** [comment]
I have merged the back into my branch. This returned that it was already up to date.

**Comment by NicoRenaud** [comment]
did you update the master branch on your machine first ?

**Comment by Rik-Breebaart** [comment]
Yes, I first pulled master into my local master branch the switched to FermiNet branch and merged this with master. However this also returns up to date.
I can see that the __init__ file from my local master file has changed matching the remote master but nothing changed in FermiNet branch

**Comment by NicoRenaud** [comment]
we switched to #38
